### ISSUE TITLE
Cannot generate day-night composites

### ISSUE THREAD

**Issue by qubitqualia** [pytroll/satpy#2664]
I am trying to produce day-night composites using GOES-18 datasets but so far have been unable to do so. The code below tested 'true_color', 'night_microphysics' and 'true_color_with_night_ir_custom' composites, where the last composite was defined in abi.yaml as:

```yaml
  true_color_with_night_ir_custom:
    compositor: !!python/name:satpy.composites.DayNightCompositor
    standard_name: true_color_with_night_ir_custom
    lim_low: 80
    lim_high: 90
    prerequisites:
      - true_color
      - night_microphysics
```

**To Reproduce**

```python
from satpy import Scene
from glob import glob


# Populate filenames array
nc_path = 'nc'
filenames = glob(nc_path + '/*.nc')
print('   ------------------------------------------ nc Filenames ------------------------------------------')
for f in filenames:
    f = f.lstrip('nc\\')
    print('   ' + f)
print('   --------------------------------------------------------------------------------------------------')
print()

# Create scene and check available composites
scene1 = Scene(reader='abi_l2_nc', filenames=filenames)
comps = scene1.available_composite_names()
print('   -------------------------------------- Available Composites --------------------------------------')
for c in comps:
    print('    ' + c)
print('   --------------------------------------------------------------------------------------------------')
print()

# Create and save true color composite
scene1.load(['true_color'])
print('   ------------------------------------ True Color Loaded Datasets ----------------------------------')
for _id in scene1.keys():
    shape = scene1[_id].attrs['area'].shape
    res = scene1[_id].attrs['area'].resolution
    print('   ' + _id['name'] + '   shape: {}'.format(str(shape)) + '  resolution: {}'.format(str(res)))
print('   --------------------------------------------------------------------------------------------------')
print()

fn = 'images/true_color.png'
new_scene = scene1.resample(scene1.coarsest_area(), resampler='native')
new_scene.save_dataset('true_color', writer='simple_image', filename=fn, fill_value=False)

# Create and save night microphysics composite
scene2 = Scene(reader='abi_l2_nc', filenames=filenames)
scene2.load(['night_microphysics'])
print('   ------------------------------------ Night Microphysics Loaded Datasets --------------------------')
for _id in scene2.keys():
    shape = scene2[_id].attrs['area'].shape
    res = scene2[_id].attrs['area'].resolution
    print('   ' + _id['name'] + '   shape: {}'.format(str(shape)) + '  resolution: {}'.format(str(res)))
print('   --------------------------------------------------------------------------------------------------')
print()

fn = 'images/night_microphysics.png'
new_scene = scene2.resample(scene2.coarsest_area(), resampler='native')
new_scene.save_dataset('night_microphysics', writer='simple_image', filename=fn, fill_value=False)


# Create and save day-night composite
scene3 = Scene(reader='abi_l2_nc', filenames=filenames)
scene3.load(['true_color_with_night_ir_custom'])
print('   ------------------------------------ Day-Night Loaded Datasets -----------------------------------')
for _id in scene3.keys():
    shape = scene3[_id].attrs['area'].shape
    res = scene3[_id].attrs['area'].resolution
    print('   ' + _id['name'] + '   shape: {}'.format(str(shape)) + '  resolution: {}'.format(str(res)))
print('   --------------------------------------------------------------------------------------------------')
print()

fn = 'images/day_night.png'
new_scene = scene3.resample(scene3.coarsest_area(), resampler='native')
new_scene.save_dataset('true_color_with_night_ir_custom', writer='simple_image', filename=fn, fill_value=False)

```

**Expected behavior**
This code should produce three images: 1) true-color composite, 2) night microphysics composite and 3) a day-night composite of true-color and night microphysics

**Actual results**
![output1](https://github.com/pytroll/satpy/assets/39470632/650f5db0-3beb-4798-ba50-419be30dece7)
![output2](https://github.com/pytroll/satpy/assets/39470632/16f9d284-358e-4647-8bf4-ea633289556b)

**Note:** This problem persists no matter which day-night composite I select or how I resample the data. For example, resampling using the below statement produces the same result:

```python
new_scene = scene3.resample(scene3['night_microphysics'].attrs['area'], resampler='nearest')
```

**Screenshots**

True color image output
![true_color_output](https://github.com/pytroll/satpy/assets/39470632/9e955325-4944-4202-9cea-419ef856ca36)

Night microphysics image output
![night_microphysics_output](https://github.com/pytroll/satpy/assets/39470632/68fcc538-2ae4-4585-bf06-48e092f5653b)

**Environment Info:**
 - OS: Windows
 - Satpy Version: 0.45.0
 - PyResample Version: 1.27.1
 - Readers and writers dependencies (when relevant):

**Comment by djhoese** [comment]
This could be a misconfiguration of the composites when L2 data is loaded. Is there a reason you're trying to generate these "level 1" products from Level 2 data files? Not that that is technically a problem, but the level 1b files are what most Satpy ABI users are using when they generate these composites.

I'll have to look at using the L2 files for this.

**Comment by qubitqualia** [comment]
Confirmed that using L1b datasets with reader 'abi-l1b' produced the same result as reported above for L2.

**Comment by djhoese** [comment]
I'll try to look tomorrow. Just to make sure, what version of Python are you using? @pnuu I wonder if this is a regression after your DayNightCompositor fixes.

**Comment by qubitqualia** [comment]
Python version is 3.9.13

**Comment by pnuu** [comment]
I only have the reduced resolution version of GOES-18/ABI data from EUMETCast, and your composite definition work for me on Python 3.11.

**Comment by qubitqualia** [comment]
@pnuu Do you happen to know the resolution and/or data shape that worked? I will try to coarsen my datasets to match before feeding them into satpy. I can't access the data directly from EUMETCast. I downloaded my data from AWS.

**Comment by djhoese** [comment]
@qubitqualia I was also able to get this to work. Keep in mind I think I have the newest Satpy, new pyresample, and new trollimage (I'm pretty sure at least). I was able to get it to work both resampling to the finest (default) and the coarsest area.

```bash
$ ls -1 /data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C??*.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C01_G16_s20182531700311_e20182531711078_c20182531711121.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C02_G16_s20182531700311_e20182531711078_c20182531711113.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C03_G16_s20182531700311_e20182531711078_c20182531711123.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C04_G16_s20182531700311_e20182531711078_c20182531711104.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C05_G16_s20182531700311_e20182531711078_c20182531711122.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C06_G16_s20182531700311_e20182531711084_c20182531711114.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C07_G16_s20182531700311_e20182531711090_c20182531711124.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C08_G16_s20182531700311_e20182531711078_c20182531711124.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C09_G16_s20182531700311_e20182531711084_c20182531711147.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C10_G16_s20182531700311_e20182531711090_c20182531711143.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C11_G16_s20182531700311_e20182531711078_c20182531711148.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C12_G16_s20182531700311_e20182531711084_c20182531711143.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C13_G16_s20182531700311_e20182531711089_c20182531711150.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C14_G16_s20182531700311_e20182531711078_c20182531711150.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C15_G16_s20182531700311_e20182531711084_c20182531711149.nc
/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C16_G16_s20182531700311_e20182531711090_c20182531711149.nc
```

These were all generated from CSPP Geo GRB or maybe were downloaded from AWS. Note there all from 2018, but that shouldn't matter once.

```bash
$ python <<EOF
from satpy import Scene
from glob import glob

scn = Scene(reader="abi_l1b", filenames=glob("/data/satellite/abi/2018253/OR_ABI-L1b-RadF-M3C??*.nc"))
scn.load(["true_color_with_night_ir_custom"])
new_scn = scn.resample(scn.coarsest_area(), resampler="native")
new_scn.save_datasets()
EOF
```

I wonder if your multiple Scenes in your script is some how getting something cached or reused that we don't want. Could you try only creating the final Scene and doing the equivalent of my script above?

**Comment by qubitqualia** [comment]
@djhoese I figured out the problem...well, roughly, anyways. At some point while trying to produce composites with Satpy I encountered an error related to the Pandas get_loc function which I found an easy solution for was to roll back Pandas to version 1.5.0. Everything then worked fine, except of course, for the day-night compositing. So, after performing the test you mentioned above and getting the same result I got previously, I started digging into all of my package versions. I upgraded Pandas to 2.1.3 but still could not create a day-night composite. However, I next upgraded xarray from 0.19.0 to 2023.11.0 and then I could could make a day-night composite using the following resample method:

```python
new_scene = scene.resample(scene.coarsest_area(), resampler='native')
```

![day_night2](https://github.com/pytroll/satpy/assets/39470632/8eaba908-9f23-47b6-979c-26772c744dc2)



Anyways, the bottom line for anyone having issues with day-night composites is to first make sure your xarray and pandas are up-to-date.
